### PR TITLE
[Typed throws] Implement reabstraction thunks that change the error

### PR DESCRIPTION
**Pull request by DougGregor**
Introduce SILGen support for reabstractions thunks that change the error, between indirect and direct errors as well as conversions amongst error types (e.g., from concrete to `any Error`).

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please build toolchain

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please build toolchain

**Comment by DougGregor** [comment]
macOS toolchain failure is unrelated. I have more fixes coming, but this is a good step forward---one can write and use a basic `map` implementation using typed throws.

File: test/IRGen/typed_throws.swift [hunk 1]
```diff
@@ -30,8 +32,36 @@ func buildMetatype() -> Any.Type {
 }
 
 // CHECK-NOMANGLE: define linkonce_odr hidden swiftcc %swift.metadata_response @"$sySi12typed_throws10MyBigErrorOYKcMa"
-// CHECK: call ptr @swift_getExtendedFunctionTypeMetadata({{i32|i64}} 2231369729, {{i32|i64}} 0, ptr [[PARAMS:%.*]], ptr null, ptr getelementptr inbounds (%swift.full_existential_type, ptr @"$sytN", i32 0, i32 1), ptr null, i32 1, ptr getelementptr inbounds (<{ ptr, ptr, i64, ptr }>, ptr {{.*}}@"$s12typed_throws10MyBigErrorOMf"
+// CHECK-NOMANGLE: call ptr @swift_getExtendedFunctionTypeMetadata({{i32|i64}} 2231369729, {{i32|i64}} 0, ptr [[PARAMS:%.*]], ptr null, ptr getelementptr inbounds (%swift.full_existential_type, ptr @"$sytN", i32 0, i32 1), ptr null, i32 1, ptr getelementptr inbounds (<{ ptr, ptr, i64, ptr }>, ptr {{.*}}@"$s12typed_throws10MyBigErrorOMf"
```
Comment (finagolfin) [inline test/IRGen/typed_throws.swift:35]: @DougGregor, this test change [fails on the Android armv7 CI](https://ci-external.swift.org/job/oss-swift-RA-linux-ubuntu-16.04-android/5804/consoleText), probably on other 32-bit platforms like watchOS or embedded too:
```
/home/ubuntu/jenkins/workspace/oss-swift-RA-linux-ubuntu-16.04-android/swift/test/IRGen/typed_throws.swift:35:20: error: CHECK-NOMANGLE: expected string not found in input
// CHECK-NOMANGLE: call ptr @swift_getExtendedFunctionTypeMetadata({{i32|i64}} 2231369729, {{i32|i64}} 0, ptr [[PARAMS:%.*]], ptr null, ptr getelementptr inbounds (%swift.full_existential_type, ptr @"$sytN", i32 0, i32 1), ptr null, i32 1, ptr getelementptr inbounds (<{ ptr, ptr, i64, ptr }>, ptr {{.*}}@"$s12typed_throws10MyBigErrorOMf"
                   ^
```

Comment (DougGregor) [reply test/IRGen/typed_throws.swift:35]: Oh, thank you for pointing this out! I'll make the check more general

Comment (DougGregor) [reply test/IRGen/typed_throws.swift:35]: Should be fixed by https://github.com/apple/swift/pull/69792

**Pull request merged**